### PR TITLE
get and update technology handles for the sdk

### PR DESCRIPTION
**Pull request by frisbee09**

### CODE REVIEW

File: src/origin_sdk/gql/queries/input_queries/session.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    return f'''query ( $sessionId: String!, {", ".join([f"""${
+        key
+    }: {
+        value
+    }""" for key, value in variables.items()])}) {
+        tree_to_string(query)
+    }'''
+
+
+config_tech_params_subtree = {
```
Comment (OscarPeace) [inline src/origin_sdk/gql/queries/input_queries/session.py:27]: Is there any way we can query the GQL schema to populate this set of fields? I feel like this is going to be another thing to maintain

File: src/origin_sdk/gql/queries/input_queries/utils.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+from typing import Dict, Union
+
+RecursiveTree = Dict[str, Union["RecursiveTree", None, str, bool, list]]
+
+
+def get_new_line(level: int):
+    return "\n" + "\t" * level
```
Comment (OscarPeace) [inline src/origin_sdk/gql/queries/input_queries/utils.py:7]: Is the indentation important?

File: src/origin_sdk/gql/queries/input_queries/utils.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                ],
+            ]
+        )
+        + f"{get_new_line(level-1)}}}"
+    )
+
+
+transform_fields: RecursiveTree = {"type": None, "value": None}
+
+transform_fields_with_cause: RecursiveTree = {
```
Comment (OscarPeace) [inline src/origin_sdk/gql/queries/input_queries/utils.py:31]: This looks unused. I think I'd want to see it in variable_values_with_transforms

File: src/origin_sdk/service/Project.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        self.session = session
+        self.project = self.session.get_project(project_id)
+
+    def get(self, key: str):
+        """
+        Shortcut for Project.project.get()
+        """
+        return self.project.get(key)
+
+    def pin(self):
```
Comment (OscarPeace) [inline src/origin_sdk/service/Project.py:23]: Feels like a weird thing to want to script.

**Comment by OscarPeace** [comment]
I think the docs are a bit out of date. https://auroraenergyresearch.github.io/aurora-origin-python-sdk/docs/category/sdk-reference
I'd like to see some of the more complicated technology queries with lots of variables in the common patterns.
The code is overall very readable and I think it will be easy to extend to include new fields and easy to copy for new datagroups. 
I question how we'll deploy the changes out to users and if we can do it seamlessly enough to not be a nuisance when we add a new plant field for example.

**Comment by OscarPeace** [comment]
I can't 100% verify the code matches the GQL schema is all cases but I hope you've tested it

**Review by OscarPeace — APPROVED** [review]

File: src/origin_sdk/gql/queries/input_queries/session.py (hunk 1 above)
Comment (frisbee09) [reply src/origin_sdk/gql/queries/input_queries/session.py:27]: Yeah, I agree with the concern. I reckon we could roll forward with this and see what the overhead is to maintain this, though.

The only fields we currently _need_ here are `name`, `appliesToEndoBui` and `appliesToExo`. The queried field is an array, and all configured parameters will come back. So there is a small degree of maintenance that automatically happens. All the other fields are not important for the SDK at the moment.

The other fields are support for UI functionality that I'm not (as of now) trying to reimplement.

File: src/origin_sdk/gql/queries/input_queries/utils.py (hunk 2 above)
Comment (frisbee09) [reply src/origin_sdk/gql/queries/input_queries/utils.py:7]: Not really, it just makes it pretty.

File: src/origin_sdk/gql/queries/input_queries/utils.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from typing import Dict, Union
+
+RecursiveTree = Dict[str, Union["RecursiveTree", None, str, bool, list]]
+
+
+def get_new_line(level: int):
+    return "\n" + "\t" * level
+
+
+def tree_to_string(tree: RecursiveTree, level=1):
```
[the remark this answers is not in the document]
Comment (frisbee09) [reply src/origin_sdk/gql/queries/input_queries/utils.py:10]: I wanted a better structure to define GQL queries. A dictionary implementation of a tree works well, and allowing it to recurse also works nicely.

What we have here deals with all the issues I have with the approach we took in the UI.

File: src/origin_sdk/gql/queries/input_queries/utils.py (hunk 3 above)
Comment (frisbee09) [reply src/origin_sdk/gql/queries/input_queries/utils.py:31]: I am copying the fragments we defined in the UI, I will double check I haven't missed something.

File: tests/test_gql_utils.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+test_object: RecursiveTree = {
+    "a": None,
+    "b": None,
+    "c": "ShouldNotTree",
+    "d": ["Could", "Tree", "But", "shouldn't"],
+    "e": {"Will": True, "For": True, "Sure": True, "Tree": False},
+    "getTechnology (technologyName: $technologyName)": {"name": None},
+}
```
Comment (frisbee09) [inline tests/test_gql_utils.py:14]: Here are the tests for the `tree_to_string` functionality we were talking about this morning.

**Pull request merged**